### PR TITLE
Codegen advance_aggregates for SUM transition function

### PR DESCRIPTION
**Pull request by armenatzoglou**
In this PR, we generate code for advance_aggregates. In particular, we create  AdvanceAggregateGenerator and enroll for advance_aggregates. 

We support codegen for SUM function on int4 and float8 datatypes. We don't support percentile & ordered aggregate. Also, we do not support null and pass-by-ref arguments. 

**Test**
`set codegen=on;`
`set client_min_messages=debug1;`
`create table foo (i float8, j float8);`
`insert into foo values (0.1, 0.2), (0.1, 0.2);`
`select sum(i) from foo;`
`select sum(i) from foo group by j;` 
`select sum(i*j + i) from foo;`

### CODE REVIEW

**Comment by armenatzoglou** [comment]
@hardikar  @foyzur @karthijrk: Please take a look if you have some time. Thanks

File: src/backend/codegen/advance_aggregates_codegen.cc [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+  llvm::Value *llvm_tuplecontext = codegen_utils->GetConstant<MemoryContext>(
+      aggstate_->tmpcontext->ecxt_per_tuple_memory);
+
+  irb->SetInsertPoint(entry_block);
+
+#ifdef CODEGEN_DEBUG
+  codegen_utils->CreateElog(DEBUG1, "Codegen'ed advance_aggregates called!");
+#endif
+
+  // Temporary variables that replace the use of slot and FunctionInfoData
```
Comment (hardikar) [inline src/backend/codegen/advance_aggregates_codegen.cc]: variables.
I think we should elaborate on this comment a bit more.

File: src/backend/codegen/advance_aggregates_codegen.cc [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+  llvm::Value *llvm_arg = irb->CreateAlloca(
```
Comment (hardikar) [inline src/backend/codegen/advance_aggregates_codegen.cc]: 1. the second argument is of type Value*, so you should replace this will nullptr
2. the third argument is the name of the instruction (variable name in LLVM) - which we don't really set anywhere else - that we probably could for better debuggability...

File: src/backend/codegen/include/codegen/advance_aggregates_codegen.h [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+   * This implementation does not support percentile and ordered aggregates.
+   *
+   * If at execution time, we see any of the above types of attributes,
+   * we fall backs to the regular function.
+   *
+   */
+  bool GenerateCodeInternal(gpcodegen::GpCodegenUtils* codegen_utils) final;
+
+ private:
+  AggState *aggstate_;
```
Comment (karthijrk) [inline src/backend/codegen/include/codegen/advance_aggregates_codegen.h]: Minor : agg_state_?

File: src/backend/executor/execQual.c [hunk 4]
```diff
@@ -5984,7 +5984,7 @@ ExecCleanTargetListLength(List *targetlist)
  * of *isDone = ExprMultipleResult signifies a set element, and a return
  * of *isDone = ExprEndResult signifies end of the set of tuple.
  */
-static bool
+bool
```
Comment (karthijrk) [inline src/backend/executor/execQual.c]: ifdef USE_CODEGEN  ?

File: src/backend/executor/nodeAgg.c [hunk 5]
```diff
@@ -1400,7 +1400,7 @@ agg_retrieve_direct(AggState *aggstate)
 							{
 								has_partial_agg = true;
 								tmpcontext->ecxt_outertuple = outerslot;
-								advance_aggregates(aggstate, pergroup, &(aggstate->mem_manager));
+								call_AdvanceAggregates(aggstate, pergroup, &(aggstate->mem_manager));
```
Comment (karthijrk) [inline src/backend/executor/nodeAgg.c]: Indention ?

File: src/include/codegen/codegen_wrapper.h [hunk 6]
```diff
@@ -62,7 +65,8 @@ typedef Datum (*SlotGetAttrFn) (struct TupleTableSlot *slot, int attnum, bool *i
 #define init_codegen()
 #define call_ExecVariableList(projInfo, values, isnull) ExecVariableList(projInfo, values, isnull)
 #define enroll_ExecVariableList_codegen(regular_func, ptr_to_chosen_func, proj_info, slot)
-
+#define #define call_AdvanceAggregates(aggstate, pergroup, mem_manager) advance_aggregates(aggstate, pergroup, mem_manager)
```
Comment (karthijrk) [inline src/include/codegen/codegen_wrapper.h]: typo ? It has two #define.

File: src/include/executor/executor.h [hunk 7]
```diff
@@ -301,6 +301,7 @@ extern ExprState *ExecInitExpr(Expr *node, PlanState *parent);
 extern ExprState *ExecPrepareExpr(Expr *node, EState *estate);
 extern bool ExecQual(List *qual, ExprContext *econtext, bool resultForNull);
 extern int	ExecTargetListLength(List *targetlist);
+extern bool ExecTargetList(List *targetlist, ExprContext *econtext, Datum *values, bool *isnull, ExprDoneCond *itemIsDone, ExprDoneCond *isDone);
```
Comment (karthijrk) [inline src/include/executor/executor.h]: ifdef USE_CODEGEN ?

File: src/backend/codegen/advance_aggregates_codegen.cc [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+  llvm::Function* llvm_MemoryContextSwitchTo =
+      codegen_utils->GetOrRegisterExternalFunction(MemoryContextSwitchTo,
+                                                   "MemoryContextSwitchTo");
+
+  // Function argument to advance_aggregates
+  llvm::Value* llvm_pergroup = ArgumentByPosition(advance_aggregates_func, 1);
+
+  // Generation-time constants
+  llvm::Value *llvm_tuplecontext = codegen_utils->GetConstant<MemoryContext>(
+      aggstate_->tmpcontext->ecxt_per_tuple_memory);
```
Comment (karthijrk) [inline src/backend/codegen/advance_aggregates_codegen.cc]: I am afraid of using member variable aggstate_ directly here instead of passed arguments. At least we should add a check that passed aggstate / memory context are same with what we have.

File: src/backend/codegen/advance_aggregates_codegen.cc [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+    }
+
+    int nargs = list_length(aggref->args);
+    Assert(nargs == peraggstate->numArguments);
+    Assert(peraggstate->evalproj);
+
+    if (peraggstate->evalproj->pi_isVarList)
+    {
+      llvm::Function* llvm_ExecVariableList = nullptr;
+      if (nullptr != peraggstate->evalproj->
```
Comment (karthijrk) [inline src/backend/codegen/advance_aggregates_codegen.cc]: Do we need this if / else ? Shouldn't that  peraggstate->evalproj->ExecVariableList_gen_info.ExecVariableList_fn points to right function always ?

File: src/backend/codegen/advance_aggregates_codegen.cc [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+          ExecVariableList_gen_info.code_generator) {
+        // Use the enrolled ExecVariableList version
+        llvm_ExecVariableList = codegen_utils->
```
Comment (karthijrk) [inline src/backend/codegen/advance_aggregates_codegen.cc]: Also during generation time, peraggstate->evalproj->ExecVariableList_gen_info.ExecVariableList_fn will be pointing to regular function. So taking getConstant will not switch to generated function in run time.

File: src/backend/codegen/advance_aggregates_codegen.cc [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+    irb->CreateCondBr(irb->CreateLoad(llvm_argnull),
+                      fallback_block /*true*/,
+                      advance_transition_function_block /*false*/);
+
+    // advance_transition_function block
+    // ----------
+    // We generate code for advance_transition_function.
+    irb->SetInsertPoint(advance_transition_function_block);
+
+    // advance_transition_function {{{
```
Comment (karthijrk) [inline src/backend/codegen/advance_aggregates_codegen.cc]: Minor Suggestion : Have separate private member function that generate code for advance_transition_function.

File: src/backend/codegen/advance_aggregates_codegen.cc [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+
+  // Fall back block
+  // ---------------
+  irb->SetInsertPoint(fallback_block);
+
+  codegen_utils->CreateElog(DEBUG1,
+                            "Falling back to regular advance_aggregates");
+
+  codegen_utils->CreateFallback<AdvanceAggregatesFn>(
+      codegen_utils->GetOrRegisterExternalFunction(advance_aggregates,
```
Comment (karthijrk) [inline src/backend/codegen/advance_aggregates_codegen.cc]: I am afraid of this fallback. let say we have 5 aggregates and if 3rd aggregate fallback, we end up starting agg0 for which we already computed the transvalue.

Comment (karthijrk) [inline src/backend/codegen/advance_aggregates_codegen.cc]: Fallback can happen from operator only when we have the overflow. I though we do elog::ERROR and stop an execution. Not sure we need to verify this.

File: src/backend/codegen/advance_aggregates_codegen.cc (hunk 10 above)
Comment (hardikar) [inline src/backend/codegen/advance_aggregates_codegen.cc]: +1. We should just call whatever peraggstate->evalproj->ExecVariableList_gen_info.ExecVariableList_fn that location points to. That would equate to a load + bitcast + call.

File: src/backend/codegen/advance_aggregates_codegen.cc (hunk 11 above)
Comment (hardikar) [inline src/backend/codegen/advance_aggregates_codegen.cc]: +1. It will easier to move around in the future.

File: src/backend/codegen/advance_aggregates_codegen.cc [hunk 13]
```diff
@@ -0,0 +1,264 @@
...
+    // We generate code for advance_transition_function.
+    irb->SetInsertPoint(advance_transition_function_block);
+
+    // advance_transition_function {{{
+
+    // TODO(nikos): Currently we do not support NULL attributes. However, if we
+    // support NULL attributes and transition function is strict, then we might
+    // have to initiate transValue, noTransvalue, and transValueIsNull
+    // accordingly (see invoke_agg_trans_func).
+    // In case of SUM, we do not have to support this.
```
Comment (hardikar) [inline src/backend/codegen/advance_aggregates_codegen.cc]: Would it be better to add those cases as ```if``` checks here - and return false (with some elog) to denote that we don't support it yet?

File: src/backend/codegen/advance_aggregates_codegen.cc [hunk 14]
```diff
@@ -0,0 +1,264 @@
...
+    llvm::Value *llvm_oldContext = irb->CreateCall(llvm_MemoryContextSwitchTo,
+                                                   {llvm_tuplecontext});
+
+    // Retrieve pergroup's useful members
+    llvm::Value* llvm_pergroupstate = irb->CreateGEP(
+        llvm_pergroup, {codegen_utils->GetConstant(aggno)});
+    llvm::Value* llvm_pergroupstate_transValue =
+        codegen_utils->GetPointerToMember(
+            llvm_pergroupstate, &AggStatePerGroupData::transValue);
+    llvm::Value* llvm_pergroupstate_transValueIsNull =
```
Comment (hardikar) [inline src/backend/codegen/advance_aggregates_codegen.cc]: This is a ptr not a value. By our naming convention, the name should reflect that.

File: src/backend/codegen/advance_aggregates_codegen.cc [hunk 15]
```diff
@@ -0,0 +1,264 @@
...
+    // In case of SUM, we do not have to support this.
+
+    //oldContext = MemoryContextSwitchTo(tuplecontext);
+    llvm::Value *llvm_oldContext = irb->CreateCall(llvm_MemoryContextSwitchTo,
+                                                   {llvm_tuplecontext});
+
+    // Retrieve pergroup's useful members
+    llvm::Value* llvm_pergroupstate = irb->CreateGEP(
+        llvm_pergroup, {codegen_utils->GetConstant(aggno)});
+    llvm::Value* llvm_pergroupstate_transValue =
```
Comment (hardikar) [inline src/backend/codegen/advance_aggregates_codegen.cc]: ptr.

File: src/backend/codegen/advance_aggregates_codegen.cc [hunk 16]
```diff
@@ -0,0 +1,264 @@
...
+        codegen_utils->GetPointerToMember(
+            llvm_pergroupstate, &AggStatePerGroupData::transValue);
+    llvm::Value* llvm_pergroupstate_transValueIsNull =
+        codegen_utils->GetPointerToMember(
+            llvm_pergroupstate, &AggStatePerGroupData::transValueIsNull);
+    llvm::Value* llvm_pergroupstate_noTransValue =
```
Comment (hardikar) [inline src/backend/codegen/advance_aggregates_codegen.cc]: ptr.

File: src/backend/codegen/advance_aggregates_codegen.cc [hunk 17]
```diff
@@ -0,0 +1,264 @@
...
+    PGFuncGeneratorInterface* pg_func_gen =
+        OpExprTreeGenerator::GetPGFuncGenerator(peraggstate->transfn.fn_oid);
+    if (nullptr == pg_func_gen) {
+      elog(DEBUG1, "We do not support built-in function with oid = %d",
+           peraggstate->transfn.fn_oid);
+      return false;
+    }
+
+    llvm::Value *newVal = nullptr;
+    pg_func_gen->GenerateCode(codegen_utils, pg_func_info, &newVal);
```
Comment (hardikar) [inline src/backend/codegen/advance_aggregates_codegen.cc]: Assert return = true and/or check and return false if this fails.

File: src/backend/codegen/advance_aggregates_codegen.cc (hunk 12 above)
Comment (hardikar) [inline src/backend/codegen/advance_aggregates_codegen.cc]: Perhaps we should stick to an all-or-nothing implementation like we did for ExecEvalExpr. We can check upfront on the functionality we support and don't.

File: src/backend/codegen/include/codegen/advance_aggregates_codegen.h [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+  static constexpr char kAdvanceAggregatesPrefix[] = "AdvanceAggregates";
+
+  /**
+   * @brief Generates runtime code that implements advance_aggregates.
+   *
+   * @param codegen_utils Utility to ease the code generation process.
+   * @return true on successful generation.
+   **/
+  bool GenerateAdvanceAggregates(gpcodegen::GpCodegenUtils* codegen_utils);
+
```
Comment (hardikar) [inline src/backend/codegen/include/codegen/advance_aggregates_codegen.h]: Reminder to run cpplint.py before pushing - since I've forgotten it many times :/

File: src/backend/codegen/include/codegen/op_expr_tree_generator.h [hunk 19]
```diff
@@ -45,6 +45,9 @@ class OpExprTreeGenerator : public ExprTreeGenerator {
         ExprTreeGeneratorInfo* gen_info,
         std::unique_ptr<ExprTreeGenerator>* expr_tree);
 
+  static gpcodegen::PGFuncGeneratorInterface* GetPGFuncGenerator(
```
Comment (hardikar) [inline src/backend/codegen/include/codegen/op_expr_tree_generator.h]: Add comments here.

File: src/backend/executor/execProcnode.c [hunk 20]
```diff
@@ -589,7 +589,9 @@ ExecInitNode(Plan *node, EState *estate, int eflags)
 			    AggStatePerAgg peraggstate = &aggstate->peragg[aggno];
 			    EnrollProjInfoTargetList(result, peraggstate->evalproj);
 			  }
-			}
+			  enroll_AdvanceAggregates_codegen(advance_aggregates,
```
Comment (hardikar) [inline src/backend/executor/execProcnode.c]: So, when we decided where to enroll ExecVariableList codegen, I remember we chose the spot right next to where the ProjectionInfo is initialized, since that's where we put the ExecVariableListCodegenInfo. Since this goes in aggstate, should be initialized where aggstate is ? Or we can initialize codegeninfos with the structures that contain them, and enroll code generators in execProcNode. Either method is fine, but let's be consistent about it :)

File: src/include/codegen/codegen_wrapper.h [hunk 21]
```diff
@@ -240,6 +260,11 @@ ExecEvalExprCodegenEnroll(ExecEvalExprFn regular_func_ptr,
         (ExecEvalExprFn)regular_func, (ExecEvalExprFn*)ptr_to_regular_func_ptr, exprstate, econtext, plan_state); \
         Assert(exprstate->evalfunc == regular_func); \
 
+#define enroll_AdvanceAggregates_codegen(regular_func, ptr_to_regular_func_ptr, aggstate) \
+		aggstate->AdvanceAggregates_gen_info.code_generator = AdvanceAggregatesCodegenEnroll( \
```
Comment (hardikar) [inline src/include/codegen/codegen_wrapper.h]: indentation

File: src/backend/codegen/advance_aggregates_codegen.cc (hunk 1 above)
Comment (armenatzoglou) [inline src/backend/codegen/advance_aggregates_codegen.cc]: Sure.

File: src/backend/codegen/advance_aggregates_codegen.cc (hunk 2 above)
Comment (armenatzoglou) [inline src/backend/codegen/advance_aggregates_codegen.cc]: Ok, I will keep one parameter only.

File: src/backend/codegen/include/codegen/advance_aggregates_codegen.h (hunk 3 above)
Comment (armenatzoglou) [inline src/backend/codegen/include/codegen/advance_aggregates_codegen.h]: I prefer to keep it as aggstate_, since we use "aggstate" almost everywhere.

File: src/backend/executor/execQual.c (hunk 4 above)
Comment (armenatzoglou) [inline src/backend/executor/execQual.c]: Yes

File: src/include/codegen/codegen_wrapper.h (hunk 6 above)
Comment (armenatzoglou) [inline src/include/codegen/codegen_wrapper.h]: Good catch! Thanks

File: src/include/executor/executor.h (hunk 7 above)
Comment (armenatzoglou) [inline src/include/executor/executor.h]: Yes

File: src/backend/codegen/advance_aggregates_codegen.cc (hunk 8 above)
Comment (armenatzoglou) [inline src/backend/codegen/advance_aggregates_codegen.cc]: Agree

File: src/backend/codegen/advance_aggregates_codegen.cc (hunk 12 above)
Comment (armenatzoglou) [inline src/backend/codegen/advance_aggregates_codegen.cc]: @karthijrk You are right. We should not fallback when an overflow occurs. Instead we should error out. I will fix that. However, we should fallback when an attribute is null. 

As we already discussed and as @hardikar mentioned above, this is an all-or-nothing approach. As @hardikar suggested, we can add a verification step upfront, which will iterate over all transition functions, and for every function, it will check if it can be codegened. I believe that this can be a separate story, since we need to decide how to keep this information (a hash table?). Do not forget that we cannot use the existing OpExprTreeGenerator::supported_function_ map, since a built-in function can be used by multiple aggregate functions.

File: src/backend/codegen/advance_aggregates_codegen.cc (hunk 14 above)
Comment (armenatzoglou) [inline src/backend/codegen/advance_aggregates_codegen.cc]: Will do.

File: src/backend/codegen/advance_aggregates_codegen.cc (hunk 15 above)
Comment (armenatzoglou) [inline src/backend/codegen/advance_aggregates_codegen.cc]: Will do.

File: src/backend/codegen/advance_aggregates_codegen.cc (hunk 16 above)
Comment (armenatzoglou) [inline src/backend/codegen/advance_aggregates_codegen.cc]: Will do.

File: src/backend/codegen/advance_aggregates_codegen.cc (hunk 17 above)
Comment (armenatzoglou) [inline src/backend/codegen/advance_aggregates_codegen.cc]: Agree, will do.

File: src/backend/codegen/include/codegen/advance_aggregates_codegen.h (hunk 18 above)
Comment (armenatzoglou) [inline src/backend/codegen/include/codegen/advance_aggregates_codegen.h]: :) I will do it.

File: src/backend/codegen/include/codegen/op_expr_tree_generator.h (hunk 19 above)
Comment (armenatzoglou) [inline src/backend/codegen/include/codegen/op_expr_tree_generator.h]: I will do that.

File: src/backend/executor/execProcnode.c (hunk 20 above)
Comment (armenatzoglou) [inline src/backend/executor/execProcnode.c]: Correct. In this case, we enroll advance_aggregates right after the initialization of aggstate. Please see lines 577 and 586.

File: src/backend/codegen/advance_aggregates_codegen.cc (hunk 13 above)
Comment (armenatzoglou) [inline src/backend/codegen/advance_aggregates_codegen.cc]: I agree. Good idea. Thanks

File: src/backend/executor/nodeAgg.c (hunk 5 above)
Comment (armenatzoglou) [inline src/backend/executor/nodeAgg.c]: Can you please elaborate more?

File: src/include/codegen/codegen_wrapper.h (hunk 21 above)
Comment (armenatzoglou) [inline src/include/codegen/codegen_wrapper.h]: Can you please elaborate more? I use Google C++ code style and the indentation is similar to all the rest enroll_* definitions. Thanks

File: src/backend/codegen/advance_aggregates_codegen.cc (hunk 12 above)
Comment (armenatzoglou) [inline src/backend/codegen/advance_aggregates_codegen.cc]: There is something missing in my previous comment related to the fallback case. At runtime, we should never fallback, since we have modified the transvalue of the previously executed transition functions. Actually, we can only fallback if we keep states of the transValues of all transition functions, so that we can rollback. I do not think that we need to support something like this for now, since we have a story to support nulls. Thus, for now, whenever we retrieve a null attribute we error out.

File: src/backend/executor/nodeAgg.c (hunk 5 above)
Comment (karthijrk) [inline src/backend/executor/nodeAgg.c]: Never mind. When I was looking diff in split windows it was appeared to have the wrong indention. But looks like they are good.

File: src/backend/codegen/advance_aggregates_codegen.cc (hunk 13 above)
Comment (armenatzoglou) [inline src/backend/codegen/advance_aggregates_codegen.cc]: More updates on this. This comment mentions that we do not support NULL attributes. If during execution advance_aggregates receives a NULL attribute then it will error out. I will update this comment.

File: src/backend/codegen/advance_aggregates_codegen.cc [hunk 22]
```diff
@@ -0,0 +1,297 @@
...
+
+AdvanceAggregatesCodegen::AdvanceAggregatesCodegen(
+    CodegenManager* manager,
+    AdvanceAggregatesFn regular_func_ptr,
+    AdvanceAggregatesFn* ptr_to_regular_func_ptr,
+    AggState *aggstate)
+: BaseCodegen(manager,
+              kAdvanceAggregatesPrefix,
+              regular_func_ptr,
+              ptr_to_regular_func_ptr),
```
Comment (foyzur) [inline src/backend/codegen/advance_aggregates_codegen.cc]: Isn't it ptr_to_chosen_func_ptr?

File: src/backend/codegen/slot_getattr_codegen.cc [hunk 23]
```diff
@@ -662,7 +662,8 @@ bool SlotGetAttrCodegen::GenerateSlotGetAttr(
   // Note: We collect error code information, based on the block from which we
   // fall back, and log it for debugging purposes.
   irb->SetInsertPoint(fallback_block);
-  llvm::PHINode* llvm_error = irb->CreatePHI(codegen_utils->GetType<char *>(), 2);
+  llvm::PHINode* llvm_error = irb->CreatePHI(
```
Comment (karthijrk) [inline src/backend/codegen/slot_getattr_codegen.cc]: Being in the same PR is fine with me. But please make the slot commit as separate as this is not related to advance transition function.

File: src/backend/codegen/advance_aggregates_codegen.cc [hunk 24]
```diff
@@ -0,0 +1,327 @@
...
+  // Function argument to advance_aggregates
+  llvm::Value* llvm_aggstate_arg = ArgumentByPosition(
+      advance_aggregates_func, 0);
+  llvm::Value* llvm_pergroup_arg = ArgumentByPosition(
+      advance_aggregates_func, 1);
+
+  // Generation-time constants
+  llvm::Value* llvm_aggstate = codegen_utils->GetConstant(aggstate_);
+
+  irb->SetInsertPoint(entry_block);
```
Comment (karthijrk) [inline src/backend/codegen/advance_aggregates_codegen.cc]: Do we need entry block ?

File: src/backend/codegen/advance_aggregates_codegen.cc [hunk 25]
```diff
@@ -0,0 +1,327 @@
...
+  // case that transition function is strict.
+
+  // oldContext = MemoryContextSwitchTo(tuplecontext);
+  llvm::Value *llvm_oldContext = irb->CreateCall(llvm_MemoryContextSwitchTo,
+                                                 {llvm_tuplecontext});
+
+  // Retrieve pergroup's useful members
+  llvm::Value* llvm_pergroupstate = irb->CreateGEP(
+      llvm_pergroup_arg, {codegen_utils->GetConstant(
+          sizeof(AggStatePerGroupData) * aggno)});
```
Comment (karthijrk) [inline src/backend/codegen/advance_aggregates_codegen.cc]: GEP take size in bytes ? sizeof return size in byte. Just want to check we are good here.

**Comment by karthijrk** [comment]
Other than minor comments, LGTM 👍  Good job

File: src/backend/codegen/advance_aggregates_codegen.cc (hunk 25 above)
Comment (hardikar) [inline src/backend/codegen/advance_aggregates_codegen.cc]: The size depends on type of pointer. We convert struct pointers to i8*, so here bytes works.

File: src/backend/codegen/slot_getattr_codegen.cc (hunk 23 above)
Comment (armenatzoglou) [inline src/backend/codegen/slot_getattr_codegen.cc]: Sure

File: src/backend/codegen/advance_aggregates_codegen.cc [hunk 26]
```diff
@@ -0,0 +1,327 @@
...
+              aggstate_(aggstate) {
+}
+
+bool AdvanceAggregatesCodegen::GenerateAdvanceTransitionFunction(
+    gpcodegen::GpCodegenUtils* codegen_utils,
+    llvm::Value* llvm_pergroup_arg,
+    int aggno,
+    llvm::Function* advance_aggregates_func,
+    llvm::BasicBlock* error_block,
+    llvm::Value *llvm_arg) {
```
Comment (hardikar) [inline src/backend/codegen/advance_aggregates_codegen.cc]: Ideally, I would like to reduce the number of arguments here less. We can revisit this issue as we implement count() and avg().
Some minor things though - we can pass llvm_pergroupstate (to avoid aggno); and let's rename llvm_arg to llvm_in_arg to clearly show they are input arguments.

File: src/backend/codegen/advance_aggregates_codegen.cc [hunk 27]
```diff
@@ -0,0 +1,327 @@
...
+
+    // advance_aggregate block
+    // ----------
+    // We generate code for advance_transition_function.
+    irb->SetInsertPoint(advance_aggregate_block);
+
+    AggStatePerAgg peraggstate = &aggstate_->peragg[aggno];
+
+    if (peraggstate->numSortCols > 0) {
+      elog(DEBUG1, "We don't codegen DISTINCT and/or ORDER by case");
```
Comment (hardikar) [inline src/backend/codegen/advance_aggregates_codegen.cc]: s/by/BY/

File: src/backend/codegen/advance_aggregates_codegen.cc [hunk 28]
```diff
@@ -0,0 +1,327 @@
...
+    }
+
+    Aggref *aggref = peraggstate->aggref;
+    if (!aggref) {
+      elog(DEBUG1, "We don't codegen non-aggref functions");
+      return false;
+    }
+
+    int nargs = list_length(aggref->args);
+    Assert(nargs == peraggstate->numArguments);
```
Comment (hardikar) [inline src/backend/codegen/advance_aggregates_codegen.cc]: I believe we use assert() macro in the C++ code, perhaps we should move to Assert(). I'm not sure of the caveats if there indeed are any of this change.

File: src/backend/codegen/advance_aggregates_codegen.cc [hunk 29]
```diff
@@ -0,0 +1,327 @@
...
+      error_aggstate_block /* false */);
+
+  // implementation block
+  // ----------
+  irb->SetInsertPoint(implementation_block);
+
+  // Since we do not support ordered functions, we do not need to store
+  // the value of the variables, which are used as input to the aggregate
+  // function, in a slot. Instead we simply store them in a stuck variable.
+  llvm::Value *llvm_arg = irb->CreateAlloca(codegen_utils->GetType<Datum>());
```
Comment (hardikar) [inline src/backend/codegen/advance_aggregates_codegen.cc]: Should have mentioned this in the first round. These allocas return pointers, am I right? Let's rename these to llvm_arg_ptr and llvm_argnull_ptr, in that case.

**Comment by hardikar** [comment]
Minor comments about naming. Rest LGTM. This gives us a big bump in performance!!! Awesome work!

File: src/backend/codegen/advance_aggregates_codegen.cc (hunk 24 above)
Comment (armenatzoglou) [inline src/backend/codegen/advance_aggregates_codegen.cc]: I prefer to keep the entry block for consistency. The first block of the code generated function is called entry block. I will remove the aggstate_check block, so that the check will be at the entry block.

File: src/backend/codegen/advance_aggregates_codegen.cc (hunk 27 above)
Comment (armenatzoglou) [inline src/backend/codegen/advance_aggregates_codegen.cc]: I will do.

File: src/backend/codegen/advance_aggregates_codegen.cc (hunk 28 above)
Comment (armenatzoglou) [inline src/backend/codegen/advance_aggregates_codegen.cc]: Good catch! Thanks

File: src/backend/codegen/advance_aggregates_codegen.cc (hunk 29 above)
Comment (armenatzoglou) [inline src/backend/codegen/advance_aggregates_codegen.cc]: Correct, I will change them.

**Comment by armenatzoglou** [comment]
Merged 875217154a8309c188b1e4b5ab08968db0560258. Thanks a lot for the code review.

File: src/backend/codegen/advance_aggregates_codegen.cc (hunk 26 above)
Comment (karthijrk) [inline src/backend/codegen/advance_aggregates_codegen.cc]: One possible way is to keep advance_aggregate_func, error_block and llvm_arg in PGFunctionGenInfo and then pass that as an argument.
